### PR TITLE
resolve relative redirect-url

### PR DESCRIPTION
**Pull request by mingwli**
resolving: https://ceapex.visualstudio.com/Engineering/_workitems/edit/91948

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4743)

### CODE REVIEW

File: src/docfx/build/redirection/RedirectionMap.cs [hunk 1]
```diff
@@ -76,16 +74,32 @@ void AddRedirections(Dictionary<string, SourceInfo<string>> items, bool checkRed
...
+
+                    Document redirect = Document.Create(docset, pathToDocset, redirectUrl);
+                    if (redirectDocumentId)
                     {
-                        errors.Add(Errors.RedirectionUrlConflict(redirectUrl));
+                        var linkType = UrlUtility.GetLinkType(redirectUrl);
+                        string mutableRedirectUrl = redirectUrl;
+                        if (linkType == LinkType.RelativePath)
+                        {
+                            mutableRedirectUrl = PathUtility.NormalizeFile(Path.Combine(Path.GetDirectoryName(redirect.SiteUrl), redirectUrl));
```
Comment (OsmondJiang) [inline src/docfx/build/redirection/RedirectionMap.cs]: Dependency.Resolve

Comment (mingwli) [reply src/docfx/build/redirection/RedirectionMap.cs]: The `DependencyResolver` is initialized inside `context`, after docset is created.

File: src/docfx/build/redirection/RedirectionMap.cs [hunk 2]
```diff
@@ -76,16 +76,32 @@ void AddRedirections(Dictionary<string, SourceInfo<string>> items, bool checkRed
...
+                        }
+                        else if (linkType != LinkType.AbsolutePath)
+                        {
+                            errors.Add(Errors.RedirectionUrlInvalid(redirectUrl));
+                            continue;
+                        }
                     }
-                    else if (checkRedirectUrl && !redirectUrls.Add(redirectUrl))
+
+                    Document redirect = Document.Create(docset, pathToDocset, redirectUrl, resolveRedirectUrl: resolveRedirectUrl);
```
Comment (OsmondJiang) [inline src/docfx/build/redirection/RedirectionMap.cs]: 👍

File: src/docfx/build/redirection/RedirectionMap.cs [hunk 3]
```diff
@@ -76,16 +76,32 @@ void AddRedirections(Dictionary<string, SourceInfo<string>> items, bool checkRed
...
                     {
-                        errors.Add(Errors.RedirectionUrlInvalid(redirectUrl));
+                        var linkType = UrlUtility.GetLinkType(redirectUrl);
+                        if (linkType == LinkType.RelativePath)
+                        {
+                            resolveRedirectUrl = true;
+                        }
+                        else if (linkType != LinkType.AbsolutePath)
+                        {
+                            errors.Add(Errors.RedirectionUrlInvalid(redirectUrl));
```
Comment (OsmondJiang) [inline src/docfx/build/redirection/RedirectionMap.cs]: do we still check absolute path need start with /?

File: src/docfx/docset/Document.cs [hunk 4]
```diff
@@ -251,7 +251,12 @@ public static Document Create(Docset docset, string path, string redirectionUrl
             }
 
             var siteUrl = PathToAbsoluteUrl(sitePath, type, schema, docset.Config.Output.Json);
-            var contentType = redirectionUrl != null ? ContentType.Redirection : type;
+            var contentType = type;
+            if (redirectionUrl != null)
+            {
+                contentType = ContentType.Redirection;
+                redirectionUrl = PathUtility.Normalize(resolveRedirectUrl ? Path.Combine(Path.GetDirectoryName(siteUrl), redirectionUrl) : redirectionUrl);
```
Comment (yufeih) [inline src/docfx/docset/Document.cs]:
```suggestion
                redirectionUrl = combineRedirectUrl ? PathUtility.Normalize(Path.Combine(Path.GetDirectoryName(siteUrl), redirectionUrl) : redirectionUrl);
```

**Pull request merged**